### ISSUE TITLE
High CPU usage (32 vCPUs) - looks due to targets discovery in K8s

### ISSUE THREAD

**Issue by lorenzo-biava** [prometheus/prometheus#8014]
**What did you do?**

We're monitoring a Kubernetes cluster consisting of about 400 nodes and 4500 pods, leveraging a single Prometheus instance with 32 vCPUs (almost fully utilized, while memory hovers between 40-50Gi). The setup leverages Prometheus Operator and most of the targets come from Service Monitors definitions (shouldn't be too relevant for the issue though).
There are about 130 target pools, with a few of those that result each in a few hundreds pods to scrape (a handful can have a couple thousands pods).
Judging by the CPU profiling graph, it looks like most of the CPU is used to update those target pools.
[pprof.prometheus.samples.cpu.005.pb.gz](https://github.com/prometheus/prometheus/files/5333959/pprof.prometheus.samples.cpu.005.pb.gz)

![prof-prod1c-20201001](https://user-images.githubusercontent.com/5684713/95204042-d8ba7100-07e3-11eb-9155-f3ded77c79f8.png)

EDIT: We're experiencing the same in another cluster with much less total Pods (~1500) but way higher target pools (~450).

**What did you expect to see?**

Not exactly sure what overall CPU usage to expect for such load, but definitely not >60% of 32vCPUs for targets discovery alone.

In case this usage is expected (and provided it is indeed coming from targets discovery), I would expect to be able to set a custom interval for targets update to tune such behavior or some other ways to reduce CPU footprint.

**What did you see instead? Under which circumstances?**

32 vCPUs (almost fully utilized), >60% of which seems to be related to targets discovery.

I see about 80 of such pools taking more than 5 seconds to get synched (varying between 4 and 8 seconds).
If my understanding is correct, the sync is executed every 5 seconds (https://github.com/prometheus/prometheus/blob/bd53b5ff37ec414d0f22315f2b4050e5c0a44652/scrape/manager.go#L158).

**Environment**

* System information:

Linux 4.15.0-1093-azure x86_64

* Prometheus version:

prometheus, version 2.20.1 (branch: HEAD, revision: 983ebb4a513302315a8117932ab832815f85e3d2)
  build user:       root@7cbd4d1c15e0
  build date:       20200805-17:26:58
  go version:       go1.14.6

* Prometheus configuration file:
```
global:
  scrape_interval: 30s
  scrape_timeout: 10s
  evaluation_interval: 30s
  external_labels:
    cluster: prod1c
    prometheus: monitoring/prometheus-operator-prometheus
    prometheus_replica: prometheus-prometheus-operator-prometheus-0
alerting:
  alert_relabel_configs:
  - separator: ;
    regex: prometheus_replica
    replacement: $1
    action: labeldrop
  alertmanagers:
  - kubernetes_sd_configs:
    - role: endpoints
      namespaces:
        names:
        - monitoring
    scheme: http
    path_prefix: /
    timeout: 10s
    api_version: v1
    relabel_configs:
    - source_labels: [__meta_kubernetes_service_name]
      separator: ;
      regex: prometheus-operator-alertmanager
      replacement: $1
      action: keep
    - source_labels: [__meta_kubernetes_endpoint_port_name]
      separator: ;
      regex: web
      replacement: $1
      action: keep
rule_files:
- /etc/prometheus/rules/prometheus-prometheus-operator-prometheus-rulefiles-0/*.yaml
- /etc/prometheus/rules/prometheus-prometheus-operator-prometheus-rulefiles-1/*.yaml
scrape_configs:
- job_name: asraas-prod/ambassador-asraas-prod/0
  honor_timestamps: true
  scrape_interval: 30s
  scrape_timeout: 10s
  metrics_path: /metrics
  scheme: http
  kubernetes_sd_configs:
  - role: endpoints
    namespaces:
      names:
      - asraas-prod
  relabel_configs:
  - source_labels: [__meta_kubernetes_service_label_service]
    separator: ;
    regex: ambassador-admin
    replacement: $1
    action: keep
  - source_labels: [__meta_kubernetes_endpoint_port_name]
    separator: ;
    regex: ambassador-admin
    replacement: $1
    action: keep
  - source_labels: [__meta_kubernetes_endpoint_address_target_kind, __meta_kubernetes_endpoint_address_target_name]
    separator: ;
    regex: Node;(.*)
    target_label: node
    replacement: ${1}
    action: replace
  - source_labels: [__meta_kubernetes_endpoint_address_target_kind, __meta_kubernetes_endpoint_address_target_name]
    separator: ;
    regex: Pod;(.*)
    target_label: pod
    replacement: ${1}
    action: replace
  - source_labels: [__meta_kubernetes_namespace]
    separator: ;
    regex: (.*)
    target_label: namespace
    replacement: $1
    action: replace
  - source_labels: [__meta_kubernetes_service_name]
    separator: ;
    regex: (.*)
    target_label: service
    replacement: $1
    action: replace
  - source_labels: [__meta_kubernetes_pod_name]
    separator: ;
    regex: (.*)
    target_label: pod
    replacement: $1
    action: replace
  - source_labels: [__meta_kubernetes_service_name]
    separator: ;
    regex: (.*)
    target_label: job
    replacement: ${1}
    action: replace
  - separator: ;
    regex: (.*)
    target_label: endpoint
    replacement: ambassador-admin
    action: replace
- job_name: asraas-prod/bofa-eng-usa-400-krypton/0
  honor_timestamps: true
  scrape_interval: 30s
  scrape_timeout: 10s
  metrics_path: /metrics
  scheme: http
  kubernetes_sd_configs:
  - role: endpoints
    namespaces:
      names:
      - asraas-prod
  relabel_configs:
  - source_labels: [__meta_kubernetes_service_label_app]
    separator: ;
    regex: krypton
    replacement: $1
    action: keep
  - source_labels: [__meta_kubernetes_service_label_release]
    separator: ;
    regex: bofa-eng-usa-400
    replacement: $1
    action: keep
  - source_labels: [__meta_kubernetes_endpoint_port_name]
    separator: ;
    regex: kr-svc-http
    replacement: $1
    action: keep
  - source_labels: [__meta_kubernetes_endpoint_address_target_kind, __meta_kubernetes_endpoint_address_target_name]
    separator: ;
    regex: Node;(.*)
    target_label: node
    replacement: ${1}
    action: replace
  - source_labels: [__meta_kubernetes_endpoint_address_target_kind, __meta_kubernetes_endpoint_address_target_name]
    separator: ;
    regex: Pod;(.*)
    target_label: pod
    replacement: ${1}
    action: replace
  - source_labels: [__meta_kubernetes_namespace]
    separator: ;
    regex: (.*)
    target_label: namespace
    replacement: $1
    action: replace
  - source_labels: [__meta_kubernetes_service_name]
    separator: ;
    regex: (.*)
    target_label: service
    replacement: $1
    action: replace
  - source_labels: [__meta_kubernetes_pod_name]
    separator: ;
    regex: (.*)
    target_label: pod
    replacement: $1
    action: replace
  - source_labels: [__meta_kubernetes_service_name]
    separator: ;
    regex: (.*)
    target_label: job
    replacement: ${1}
    action: replace
  - separator: ;
    regex: (.*)
    target_label: endpoint
    replacement: kr-svc-http
    action: replace
- job_name: asraas-prod/bofa-eng-usa-400-krypton/1
  honor_timestamps: true
  scrape_interval: 10s
  scrape_timeout: 10s
  metrics_path: /metrics
  scheme: http
  kubernetes_sd_configs:
  - role: endpoints
    namespaces:
      names:
      - asraas-prod
  relabel_configs:
  - source_labels: [__meta_kubernetes_service_label_app]
    separator: ;
    regex: krypton
    replacement: $1
    action: keep
  - source_labels: [__meta_kubernetes_service_label_release]
    separator: ;
    regex: bofa-eng-usa-400
    replacement: $1
    action: keep
  - source_labels: [__meta_kubernetes_endpoint_port_name]
    separator: ;
    regex: kr-fluentd-metrics-port
    replacement: $1
    action: keep
  - source_labels: [__meta_kubernetes_endpoint_address_target_kind, __meta_kubernetes_endpoint_address_target_name]
    separator: ;
    regex: Node;(.*)
    target_label: node
    replacement: ${1}
    action: replace
  - source_labels: [__meta_kubernetes_endpoint_address_target_kind, __meta_kubernetes_endpoint_address_target_name]
    separator: ;
    regex: Pod;(.*)
    target_label: pod
    replacement: ${1}
    action: replace
  - source_labels: [__meta_kubernetes_namespace]
    separator: ;
    regex: (.*)
    target_label: namespace
    replacement: $1
    action: replace
  - source_labels: [__meta_kubernetes_service_name]
    separator: ;
    regex: (.*)
    target_label: service
    replacement: $1
    action: replace
  - source_labels: [__meta_kubernetes_pod_name]
    separator: ;
    regex: (.*)
    target_label: pod
    replacement: $1
    action: replace
  - source_labels: [__meta_kubernetes_service_name]
    separator: ;
    regex: (.*)
    target_label: job
    replacement: ${1}
    action: replace
  - separator: ;
    regex: (.*)
    target_label: endpoint
    replacement: kr-fluentd-metrics-port
    action: replace
[...]
```
PS: I can provide the full configuration if that's helpful, though it's quite longer

**Comment by lorenzo-biava** [comment]
Removing a few jobs that weren't scraping any targets (~20) did actually reduce the CPU load significantly (~3 vCPU)
![image](https://user-images.githubusercontent.com/5684713/95429284-75971e80-094a-11eb-9edb-3f97a85afd9a.png)

**Comment by roidelapluie** [comment]
Can we have the startup logs?

**Comment by lorenzo-biava** [comment]
Here's the log from the past few days, including startup (I'm seeing a lot of K8s API errors, which I believe are related to transient instability of that managed cluster).
[prod1c-prom-logs.zip](https://github.com/prometheus/prometheus/files/5346281/prod1c-prom-logs.zip)


This is a fresh start with debug level.
[prod1c-prom-logs-debug.zip](https://github.com/prometheus/prometheus/files/5346272/prod1c-prom-logs-debug.zip)

**Comment by roidelapluie** [comment]
You are opening 28 different kubernetes connections. You could try to reuse the exact same service discoveries config and apply relabling per job to select which targets to monitor.

cat prod1c-prom-logs-debug.txt |grep subs |grep kuber -c
28

**Comment by roidelapluie** [comment]
Thanks for the logs anyway, I will try to have a look at the code as well.

**Comment by lorenzo-biava** [comment]
@roidelapluie thanks for the suggestion; can you point me to an example of such config? I then need to check if that's possible via PromOp ServiceMonitors or something similar...
I know the overall config could be way more efficient, though we're offering the Prometheus environment to multiple teams, each of which should be able to deploy the configuration needed.

**Comment by brian-brazil** [comment]
Basically if the bit under `kubernetes_sd_configs:` is identical then we'll only instantiate one k8s discovery.

**Comment by lorenzo-biava** [comment]
So basically I need to get rid of the namespace in the `kubernetes_sd_configs` and add a `relabel_configs` to filter the namespace.

I believe that's not (currently) possible with ServiceMonitors, not even with a more general `matchExpressions` (which is limited to the Service's label: 
https://github.com/prometheus-operator/prometheus-operator/blob/v0.42.1/pkg/prometheus/promcfg.go#L911).

I'll ask around and possibly reach out to PromOp to see if there are ways to do that. However having the namespace filter in the `kubernetes_sd_configs` seemed more logical, and I was under the impression to also be more efficient (not trying to filter among all the clusters' pods).
I still hope you can find some other optimizations in the discovery code or Prometheus-wide settings to leverage 😃

**Comment by brian-brazil** [comment]
What's more efficient would depend on how many namespaces there are, and what proportion of them that this Prometheus is interested in. I'd expect specifying a namespace to always be more efficient though, so if this is the issue this smells like a performance issue in the k8s client code somewhere as the namespace filtering isn't happening properly on the server side.

**Comment by djsly** [comment]
@brian-brazil  we have 

```
❯ k get ns --no-headers | wc -l
     112
```

and 

```
❯ k get servicemonitor -A --no-headers  | awk '{ print $1}' | sort | uniq -c | sort -r
 118 asraas-stage
  87 asraas-qa
  54 gatekeeper-qa
  29 gatekeeper-dev
  24 mix-dev
  23 monitoring
  23 mix-stage
  23 mix-qa
  18 nluaas-stage
  16 gatekeeper2-dev
  16 asraas-dev
  15 nluaas-qa
   9 gatekeeper-stage
   8 fabric-stage
   8 fabric-qa
   8 fabric-dev
   8 fabric-ctp-dev
   7 fabric-ctp-qa
   6 ttsaas-dev
   6 nluaas-ctp-dev
   6 dlgaas-dev
   5 nluaas-dev
   4 ttsaas-stage
   4 svc-ps-dev
   4 nluaas-ctp-qa
   4 global-auth-stage
   4 global-auth-qa
   4 global-auth-dev
   4 global-auth-ctp-qa
   4 global-auth-ctp-dev
   4 dlgaas-qa
   3 ttsaas-qa
   3 mixidp-auth-dev
   3 media-manager-dev
   3 dlgaas-stage
   2 mixidp-auth-stage
   2 mixidp-auth-qa
   1 xarch-dev
   1 ingress-controllers
   1 cert-manager
```

**Comment by djsly** [comment]
Any insights here would be helpful. we are getting to 500 nodes with > 2000 pods and right now we would need to get > 32 CPU for prometheus.

**Comment by chancez** [comment]
> What's more efficient would depend on how many namespaces there are, and what proportion of them that this Prometheus is interested in. I'd expect specifying a namespace to always be more efficient though, so if this is the issue this smells like a performance issue in the k8s client code somewhere as the namespace filtering isn't happening properly on the server side.

There is no namespace level filtering on the server-side that I'm aware of. There's two ways you can implement it. 1 'watch all namespaces' watcher, with client side filtering, or open 1 watch per-namespace (with http2 that can still be 1 connection if you re-use the client I believe).

**Comment by lorenzo-biava** [comment]
So we managed to merge together a lot of ServiceMonitors. Even though they were all scraping the same namespace, we got an impressive reduction in CPU usage (less than half; see below).

I think this indicates the number of connections to Kubernetes ([as suggested previously](https://github.com/prometheus/prometheus/issues/8014#issuecomment-705418772)) might not be the primary contributor, while the sheer number of jobs that use K8s Service Discovery is.

Just to reiterate on this: is my assumption the service discovery for each target pool runs every 5 seconds correct ([see here](https://github.com/prometheus/prometheus/blob/bd53b5ff37ec414d0f22315f2b4050e5c0a44652/scrape/manager.go#L158))?

![image](https://user-images.githubusercontent.com/5684713/98916329-56505b80-24cb-11eb-8cb8-6f291909e532.png)

**Comment by brian-brazil** [comment]
> Just to reiterate on this: is my assumption the service discovery for each target pool runs every 5 seconds correct (see here)?

No, that's a throttle so we don't process updates from an SD more than every 5s. Processing updates is considered to be cheap, but not that cheap.

**Comment by w4rgrum** [comment]
We are currently facing the same issue on our Prometheis instances: we have ~30 kube sd jobs (pod type) that are not constrained by namespaces and there are a lot of pods running on the platform (~5K pods) and the CPU usage for those instances is abnormally high.
Having a look at the discovery page of the ui we can see that most of the jobs are keeping less than 100 targets out of ~28K each.
In order to mitigate this we added selectors to some of the kube sd jobs, which helped reduce the CPU usage by half. However this is not perfect since it seems when a selector is matching nothing you see 0/28K in the discovery which is weird.
Our impressions on this is that the less targets discovered you have as a result of the kube calls the less CPU is consumed in the end, meaning the consumption might not be linked to kube calls but to what is done afterwards (relabeling and such?).

**Comment by brian-brazil** [comment]
Can you get a CPU profile?

**Comment by w4rgrum** [comment]
CPU profile with the selectors
[profile_with_selectors.zip](https://github.com/prometheus/prometheus/files/5589800/profile_with_selectors.zip)
![profile_with_selectors](https://user-images.githubusercontent.com/54102154/100094767-cbbc1480-2e59-11eb-84d9-87dc40daf5f6.png)

CPU profile without selectors
[profile.zip](https://github.com/prometheus/prometheus/files/5589830/profile.zip)
![profile](https://user-images.githubusercontent.com/54102154/100094790-d5de1300-2e59-11eb-9f20-fc1d0edbd2bd.png)

**Comment by brian-brazil** [comment]
Looks like it's that part of the code alright. It's not something we've ever really optimised, as it's not meant to be run particularly often. k8s sending 28k targets for each of 30 scrape configs every 5 seconds wasn't exactly in mind. There's probably some low hanging fruit here.

**Comment by vitkovskii** [comment]
@brian-brazil Hey there! What if we have a label filter in the k8s discovery plugin? In our case discovery produces ~120 labels per target, but we use only 5 of them. Yes, there is a relabel config, but the main performance issue is to alloc memory for the number of labels (30000 targets * 120 labels=~3 million labels), sort them, hash for deduplication, and then send to relabeling. This long and hard work happens each scrape pool sync. What do you think?

**Comment by brian-brazil** [comment]
That sounds like slowly reinventing relabelling, so would ultimately end up with the same performance costs. We have to allocate  the labels one way or the other.

I'd suggest looking at ways to make what we have more efficient, rather than immediately jumping to adding yet more configuration for users. There's likely quite a few low hanging fruit performance wise, as the relevant code paths have never really been optimised.

**Comment by d-ulyanov** [comment]
The option from @vitkovskii sounds really reasonable. Unfortunately, we just wasting CPU and memory for useless work here and I have no idea what to do next because the number of targets in our K8S growing really rapidly. It seems that Prometheus becoming unscalable here. Let's find an engineering solution, colleagues.

**Comment by roidelapluie** [comment]
cc @simonpasquier @brancz

**Comment by vitkovskii** [comment]
Actually, this hard work is doing per service monitor. So if you have 30 of them, you have x30 useless work. Let's count: 120 labels is about 6Kb, 30000 targets, and 30 service monitors. It's about 5.5Gb of RAM only to create raw label sets from discovery.

**Comment by roidelapluie** [comment]
Do you have concrete exemples about how we could avoid that? Is that e.g. underlying labels you could delete from your pods?

**Comment by brian-brazil** [comment]
> Let's count: 120 labels is about 6Kb, 30000 targets, and 30 service monitors. It's about 5.5Gb of RAM only to create raw label sets from discovery.

That'd only be if none of them were dropped, which is unlikely to be the case. We should only be keeping dropped targets once, not 30x - if not that sounds like a low hanging fruit that could be tackled.

**Comment by lorenzo-biava** [comment]
> Actually, this hard work is doing per service monitor. So if you have 30 of them, you have x30 useless work.

It's definitely amplified by the number of Service Monitors. We replaced tens of SM with a single one (for a particular application), while the overall pods/labels stayed the same (or even slightly increased) and the CPU usage got to less than 50% it was before (see https://github.com/prometheus/prometheus/issues/8014#issuecomment-725935652)

**Comment by brian-brazil** [comment]
> and the CPU usage got to less than 50% it was before

I'm confused here. https://github.com/prometheus/prometheus/issues/8014#issuecomment-737328100 is talking about RAM, and now you're talking about CPU. Which is the problem?

**Comment by djsly** [comment]
@brian-brazil  you URL doesn't pin tto the same comment as the one from @lorenzo-biava comment's 

https://github.com/prometheus/prometheus/issues/8014#issuecomment-725935652

**Comment by lorenzo-biava** [comment]
For us it was definitely CPU (as per the title of the issue). Not sure about RAM, haven't seen any correlation with a change in memory usage yet.
PS: All the provided profiles are for CPU. Let me know if you also need a memory one.

**Comment by brian-brazil** [comment]
Let's just stick with CPU then. If it is also memory that's a separate thing which we can look at optimising.

My first though on how to handle this generally so it benefits all SDs/users would be to see if we can avoid re-process a target that hasn't changed when there's refresh from SD, thus avoiding a large chunk of the relabelling processing.

**Comment by brian-brazil** [comment]
Yeah, looking at the code at https://github.com/prometheus/prometheus/blob/master/scrape/scrape.go#L417-L431 if we made droppedTargets a map we could check if an identical target was dropped last time and fastpath that, bypassing all the relabelling.

**Comment by vitkovskii** [comment]
@brian-brazil Heaviest function is `targetsFromGroup()` https://github.com/prometheus/prometheus/blob/master/scrape/scrape.go#L419. This function builds two label sets: the original and the relabeled one. So we can't identify the target before `targetsFromGroup()` because the target is identifying with its `hash(label set)`. Yes, there is some room for small optimizations around this place, but it won't let us reduce RAM and CPU usage dramatically without refactoring a large amount of code.

**Comment by vitkovskii** [comment]
Do you know what we can use better than `hash()` for a map key to skip `targetsFromGroup()`?

**Comment by brian-brazil** [comment]
My point was that we can avoid doing that work if we've previously dropped that target, as the metadata alone determines if it'll be dropped - we don't need to run it every time.

Also looking at your scrape configs, they look like they can be optimised and collapsed down into a single scrape config. The keeps can all be done in one action, and the endpoint label is always the same as the port name. The first pod action is also redundant, as it's always overwritten by the second. You can also remove the service label, as that information is already in the job label - you don't need it twice, that just clutters up your target labels.

**Comment by brian-brazil** [comment]
> Do you know what we can use better than hash() for a map key to skip targetsFromGroup()?

It's either that, or render the entire metadata string. Collisions seem unlikely given the numbers we're dealing with, so hash should be fine.

**Comment by vitkovskii** [comment]
@brian-brazil https://github.com/prometheus/prometheus/pull/8250 check it out. Did you mean something like that?

**Comment by brian-brazil** [comment]
That's in the right direction, I've commented on the PR on how to improve it.

**Comment by vitkovskii** [comment]
@brian-brazil @roidelapluie Hi guys. Here are some performance graphs:

![image](https://user-images.githubusercontent.com/2269880/101487288-000df580-396f-11eb-8440-ebce314679d3.png)
![image](https://user-images.githubusercontent.com/2269880/101487404-2d5aa380-396f-11eb-8351-3ed85a58e9c5.png)

After the first purple annotation is the #8250 fix. After second is the fix where unused labels are filtered out.

Our setup is 6 sharded Prometheus instances + 6 replicas. As you see the first optimization reduces only ~10 CPU cores, but the second ~80 cores and it's 3x lower than the original core consumption. 

Here is a label list which are going from SD and what labels we need: https://gist.github.com/vitkovskii/abc0ef5f2805165b943602aceed3dd34

**Comment by brian-brazil** [comment]
That's a big difference but as I've said above, let's not jump to optimisations that require user configuration that only work for one SD in only particular extreme use cases before we've optimised the general code path in ways that'll Just Work for everyone.

**Comment by vitkovskii** [comment]
My suggestion would change k8s discovery default behavior to discover only some basic labels and also make it possible to turn on other labels using an inclusive list or something like that.

**Comment by vitkovskii** [comment]
This problem hits any setup with k8s discovery and service monitors :) I think many engineers don't even guess that they heat the air by using Prometheus to monitor k8s.

**Comment by roidelapluie** [comment]
> This problem hits any setup with k8s discovery and service monitors :) I think many engineers don't even guess that they heat the air by using Prometheus to monitor k8s.

What does it look like by just removing the labelpresent and annotationpresent labels?

**Comment by brian-brazil** [comment]
> My suggestion would change k8s discovery default behavior to discover only some basic labels and also make it possible to turn on other labels using an inclusive list or something like that.

That's not the paradigm we take with SD, we expose everything that could be of use. That'd also be a breaking change for many users.

**Comment by brancz** [comment]
Could we not predict the source labels that relabeling rules will use, and only produce those?

**Comment by vitkovskii** [comment]
Yes, I thought about the same idea and I have a concern. Will a low amount of labels enough to identify the target using hash(). Should we mark some labels as "identifying" and don't remove them in any case?

**Comment by brian-brazil** [comment]
I don't think that matters. If there's two targets from SD whose metadata labels differ in a way that doesn't affect relabelling such that they end up with the same target labels, they were going to get de-duped anyway after relabelling.

There's two bigger issues. The first is that SDs are shared across multiple scrape configs, the second is that we expose all labels in the API/UI and silently dropping that would make both configuration and debugging harder.

**Comment by vitkovskii** [comment]
@brian-brazil Is it a strict rule that we should expose all original labels( to UI/API) in a sorted way? Or can we break the sequence? My experiments show that the most expensive operation is sorting labels. But for relabeling it isn't required labels to be sorted. So we can sort much fewer labels only after relabeling. The drawback is that the original label set will be exposed unsorted.

**Comment by roidelapluie** [comment]
> @brian-brazil Is it a strict rule that we should expose all original labels(to UI/API) in a sorted way? Or can we break the sequence? My experiments show that the most expensive operation is sorting labels. But for relabeling it isn't required labels to be sorted. So we can sort much fewer labels only after relabeling. The drawback is that the original label set will be exposed unsorted.

We could sort them in the UI, if that is really making a big difference. But I think that relabeling might somwhow have dependencies on order, alonside other sanity checks we do (like checking for duplicate labels)

**Comment by brian-brazil** [comment]
> Is it a strict rule that we should expose all original labels(to UI/API) in a sorted way? 

Yes, that's part of our documented API that cannot be broken - though the sorting doesn't matter, it is a JSON map which has no ordering.

> My experiments show that the most expensive operation is sorting labels. 

For the data structure to work it needs to be sorted, and that includes for relabelling so it can find the labels. Are we sorting more than we need to? We should only need to do it once per SD target.

**Comment by vitkovskii** [comment]
The first one is here: https://github.com/prometheus/prometheus/blob/master/scrape/target.go#L368 actually, it's not necessary for relabeling:

**Comment by vitkovskii** [comment]
>and that includes for relabelling so it can find the labels
Relabeling scans all labels and it's not required them to be sorted.

**Comment by vitkovskii** [comment]
Why relabeling take a list of labels, not a map? For every target on every config iteration, it allocates a new list. This is the second bottleneck after sorting.

**Comment by brian-brazil** [comment]
The map is the old data structure, Prometheus 2.x introduced the list for performance reasons. The map should be considered legacy, and remaining uses removed where practical.

Why not do the sort up in SD? We only need it once, not for every scrape config and it'd further prune where the old data structure is used.

> Relabeling scans all labels and it's not required them to be sorted.

It's using a library which expects them to be sorted, so that invariant should be maintained. Relabelling does require them to be sorted due to this.

**Comment by shaikatz** [comment]
@brian-brazil there are any plans to improve that area? it's still a pain for anybody who uses many service monitors in his cluster, the savings potential here is great.

**Comment by brian-brazil** [comment]
I think that's the first thing we should probably look at improving.

**Comment by brancz** [comment]
@shaikatz do you mind uploading the whole profile either as a file to github or via https://share.polarsignals.com/ ? It would be great if we could explore the profile.

I also opened https://github.com/prometheus-operator/prometheus-operator/issues/3840 as I think that is another angle from which this can be optimized on the prometheus-operator side.

**Comment by shaikatz** [comment]
@brancz 
[pprof.prometheus.samples.cpu.001.pb.gz](https://github.com/prometheus/prometheus/files/5982017/pprof.prometheus.samples.cpu.001.pb.gz)

**Comment by brancz** [comment]
Could you also share an allocs profile?

From that profile it does look to me like a lot of memory-trash is produced by discovery which causes large sweep and GC CPU usage. It does all seem to add up to https://github.com/prometheus-operator/prometheus-operator/issues/3840.

**Comment by shaikatz** [comment]
Can you provide the exact pprof command to get your required profile?

**Comment by brancz** [comment]
Yeah it is, thank you! Yeah that seems to also point in the same direction.

**Comment by lorenzo-biava** [comment]
@brancz / @brian-brazil / @roidelapluie is there any update on this issue? any options we can explore to patch this behavior?
The prom-operator's PR seems to be stuck since March...

**Comment by fenniedaddy** [comment]
Same problem here. 
The easy solution for now is to merge servicemonitor.
Try Using only one servicemonitor for all the services to be monitored, and the prometheus reload op will be more efficicent.

**Comment by wulianhuo** [comment]
Same problem here.
The easy solution for now is to merge servicemonitor.
Try Using only one servicemonitor for all the services to be monitored, and the prometheus reload op will be more efficicent.

**Comment by d-ulyanov** [comment]
As MR was not accepted we've decided to implement our own separate discovery service (we're calling it "target balancer") and deliver targets in files (fileSD) to Proms. Profits: 1) reduce kubeAPI load 2) significantly reduced CPU usage on Prom instances

**Comment by m-yosefpor** [comment]
It seems we are also hitting this issue, but the weird thing is we are hitting the issue with only 1 of our prometheus servers!! (we have configured 2 replicas in prometheus operator, no sharding). 

prometheus-0: [profile.pb.gz](https://github.com/prometheus/prometheus/files/9529550/profile.pb.gz)
<img width="1320" alt="image" src="https://user-images.githubusercontent.com/47300215/189194535-eb84db59-5be2-4f55-940b-2863701da722.png">

prometheus-1: [profile(1).pb.gz](https://github.com/prometheus/prometheus/files/9529553/profile.1.pb.gz)
<img width="1334" alt="image" src="https://user-images.githubusercontent.com/47300215/189194594-76c7d813-5ef6-4427-92e4-bbd81a1ba438.png">


 
 You can see the difference in pprof between two instances. Also the difference in CPU usage of these instances:
 
 
<img width="666" alt="image" src="https://user-images.githubusercontent.com/47300215/189194301-a586af4c-4f70-449c-86ae-14682359103a.png">

**Comment by d-ulyanov** [comment]
@m-yosefpor we finally moved out all discovery logic outside of prometheus to separate daemon and switched to simple file/http discovery, it also allowed us to implement custom sharding logic.  
Maybe it makes sense to open source this tool.

**Comment by m-yosefpor** [comment]
> @m-yosefpor we finally moved out all discovery logic outside of prometheus to separate daemon and switched to simple file/http discovery, it also allowed us to implement custom sharding logic. Maybe it makes sense to open source this tool.

We would appreciate if such tools gets open-sourced.

**Comment by iamyeka** [comment]
Waiting for a great solution, currently we set `scrape.discovery-reload-interval` to a larger duration as the temporary solution.

**Comment by iamyeka** [comment]
> It's using a library which expects them to be sorted, so that invariant should be maintained. Relabelling does require them to be sorted due to this.

I can't figure out why this could happen. Which library is being used?

**Comment by bboreham** [comment]
The problems described at the top may be improved by #12048 and #12084.

**Comment by beorn7** [comment]
Hello from the bug scrub.

We assume this problem has been improved by  #12048 and #12084 indeed. If you still see excessive CPU usage that we might be able to fix, please follow up here (or file a new issue).

**Comment by jsedy7** [comment]
Hello, I have created a question on https://discuss.prometheus.io/t/constantly-increasing-cpu-usage-on-the-prometheus-server/2304 about the increasing CPU usage of my Prometheus server.

Honestly, I'm not sure if it is related at all. Probably not, but I currently can't find a solution to my situation. Could you please advise me on which direction to take next?

![profile003](https://github.com/prometheus/prometheus/assets/8748774/fac61b92-f858-4fb3-bd08-3c51cd3c89f7)
[profile002.pb.gz](https://github.com/user-attachments/files/15859700/profile002.pb.gz)
